feat: 글 작성 시 이미지를 업로드하는 기능 추가 [ISSUE-52]

aegis1920 opened the pull request:
**이슈 번호**

resolved: #52 

**작업 내용**

1. 이미지를 AWS S3로 업로드
```
# aws.properties에 들어가는 내용

cloud.aws.credentials.accessKey=[자신의 accessKey]
cloud.aws.credentials.secretKey=[자신의 secretKey]
```
2. 이미지 확장자 검사 및 MIME 타입 검사(Apache Tika 이용)
3. Post와 Image는 일대다 양방향으로 구현
4. 외부 API(AWS S3) 테스트는 Mock 테스트로 진행 (인수테스트 포함)

**테스트 작성 여부**

- [ ] Test case

**참고 사항**

1.  AWS S3를 사용하려면 제대로 된 aws.properties를 resoureces에 추가해야 합니다.

aegis1920 committed:
feat: 이미지를 포함한 글 등록 인수테스트 작성

- 테스트용 정적 파일 추가
- PostRequest, PostResponse에 이미지를 담는 필드 추가
- Image 엔티티 작성
- Mock Data를 반환하는 이미지를 포함한 글 등록 컨트롤러 작성

aegis1920 committed:
feat: 이미지 확장자 도메인 및 테스트 작성

- NotImageExtensionException 커스텀 예외 추가
- 이미지 확장자 유효성 검사 기능 추가
- test1.txt 이름을 test4.txt로 변경

aegis1920 committed:
feat: 이미지 MIME 타입 유효성 검사 로직 및 테스트 작성

- tika 의존성 추가
- NotImageMimeTypeException 커스텀 예외 추가
- ImageService 추가
- ImageRepository 추가
- Image 엔티티에 @Entity 추가

aegis1920 committed:
feat: .gitignore에 aws.properties 추가

aegis1920 committed:
feat: AWS S3 의존성 및 설정 추가

- aws java 의존성 추가
- AWS S3 Config 추가
- application.properties bucket 명, region 명 추가

aegis1920 committed:
feat: AWS S3를 이용한 이미지 업로드 기능 추가

- S3Uploader 구현
- Controller와 Service 연결 및 테스트 수정
- ImageExtension file getter 수정
- PostRequest setter 추가, PostResponse listOf() 삭제
- ImageRepository findAllByPostId 메서드 작성
- 이미지를 포함한 포스트 생성 테스트 추가

aegis1920 committed:
feat: 테스트용 static 상수 추가 및 적용, 테스트용 FileConverter 작성

- 이미지 파일명 수정
- constants 패키지 -> utils 패키지 통합

aegis1920 committed:
feat: GlobalExceptionHandler 커스텀 예외 처리

- 상수 추가
- getName() -> getOriginalFilename() 으로 수정

Hamliet commented:
아래 4개의 테스트에서 Fail하여 CI 통과하지 못했습니다. 확인 후 커밋 다시 날려주세요~
`LegenoAroundHereApplicationTests > contextLoads() FAILED`
`acceptance.PostAcceptanceTestPostAcceptanceTest > 글 관리 FAILED`
`acceptanceTest.UserAcceptanceTestUserAcceptanceTest > manageUser() FAILED`
`controller.UserControllerTestUserControllerTest > 회원 가입 요청/응답 테스트 FAILED`

aegis1920 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#52

aegis1920 committed:
refactor: 글 작성 시 이미지 업로드 인수테스트 Mocking

aegis1920 committed:
refactor: mock aws.properties 추가 및 logback 속성 위치 변경

- .gitignore에 aws.properties 삭제
- 잘못 생성된 log 파일 삭제

aegis1920 reviewed (commented):
인수테스트도 Mocking하도록 수정했습니다!

ttkmw commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/GlobalExceptionHandler.java:
```diff
@@ -19,15 +21,17 @@
-    public ResponseEntity<ErrorResponse> HandleBadRequest(UserInputException e) {
+    @ExceptionHandler({UserInputException.class, NotImageMimeTypeException.class,
+        NotImageExtensionException.class})
+    public ResponseEntity<ErrorResponse> HandleBadRequest(IllegalArgumentException e) {
```
함수명에 카멜 케이스를 쓰는게 좋을 것 같습니다!

aegis1920 replied:
#56 PR에 이미 수정되어있어 따로 수정하지 않겠습니다 :)

ttkmw commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/GlobalExceptionHandler.java:
```diff
@@ -19,15 +21,17 @@
     @ExceptionHandler(Exception.class)
-    public ResponseEntity<ErrorResponse> handleInternalServerError() {
+    public ResponseEntity<ErrorResponse> handleInternalServerError(Exception e) {
+        e.printStackTrace();
```
콘솔에는 안찍어도 괜찮지 않을까용??

aegis1920 replied:
e.getMessage()를 남긴다는 게 e.printStackTrace()를 작성했네요;; 지적 감사합니당 👍

ttkmw commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/ImageService.java:
```diff
@@ -0,0 +1,64 @@
+                    multipartFile.getName() + " 파일은 이미지 MIME 타입이 아닙니다!");
+            }
+        } catch (IOException e) {
+            // TODO: 2020/07/28 추후 logger 적용
```
todo에 logger 적용 뿐 아니라, 다른 예외처리도 해주는 게 좋을 것 같습니다

혹시 IOException이 CheckedException이라 잡으신 거라면, 예외 전환을 해주는게 좋을 것 같습니다

aegis1920 replied:
커스텀 예외를 만들어 예외 전환을 통해  ControllerAdvice로 넘어가도록 했습니다! 조언 감사합니다 👍

ttkmw commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/ImageService.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    private boolean isNotImageMimeType(MultipartFile multipartFile) throws IOException {
+        Tika tika = new Tika();
```
tika가 먼가용??
어떤 역할을 해주는지 궁금합니다!

aegis1920 replied:
파일의 내용과 메타 데이터를 분석해줄 수 있는 툴킷입니다! 여기서는 MIME 타입을 확인하는데 쓰였습니다 :)
https://www.baeldung.com/apache-tika

ttkmw commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/PostService.java:
```diff
@@ -16,19 +18,32 @@
     public PostResponse createPost(PostRequest postRequest) {
         Post post = postRepository.save(postRequest.toPost());
-        return PostResponse.of(post);
+        List<Image> images = saveImages(postRequest, post);
```
현재는 createPost를 할 때 항상 saveImages를 하고, 그 안에서 만약에 없으면 빈 리스트를 리턴해주고 있는데요!
개인적으로
postRequest에 image가 있는지 확인하고, 있으면 이미지 포함해서 post 저장, 없으면 이미지 없이 post 저장 하는게 좋지 않을까 싶습니다.

aegis1920 replied:
현재 Image와 Post의 관계가 N:1로 다대일 관계를 이루고 있습니다. Image에 Post를 넣어야 하기 때문에 `postRepository.save()`가 먼저 실행되고 나서 `imageService.save()`가 일어나야 합니다. 

'PostRequest에 image가 있으면 이미지 포함해서 post 저장, 없으면 이미지 없이 post 저장'을 하려면 한 메서드에 `postRepository.save`()와`imageService.save()`를 넣게 될 거라고 생각됩니다. 이러면 가독성 측면에서 오히려 잘 읽히지 않을 것 같습니다. 한 곳에 넣지 않고 다르게 구현할 방법이 있을까요? Request Change를 주셔도 좋습니다 :)

aegis1920 replied:
객체 간 관계에 맞게 양방향으로 구현하면서 로직이 바뀔 것 같습니다. 조언 감사합니다 :)

ttkmw commented on back-end/legeno-around-here/src/main/resources/logback-spring.xml:
```diff
@@ -1,5 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <configuration scan="true" scanPeriod="60 seconds">
+  <property name="LOG_PATTERN"
```
이건 혹시 어떤 내용인가요?

로깅 관련 pr인 것 같은데, 로그 관련 수정이 필요하다면 새로운 pr을 보내주시는게 좋지 않을까 합니다!

aegis1920 replied:
property 태그가 아래로 내려가있어서 property가 먹히지 않는 이슈였는데요! 위로 올리기만 하면 돼서 커밋 내용에 담았습니다. PR을 날리는 기준에 대해서는 팀원 다 같이 이야기해보면 좋을 것 같아요 :)

aegis1920 replied:
회의 후 융통성있게 PR을 날리는 것으로 결정했습니다 👍

ttkmw commented on back-end/legeno-around-here/src/test/resources/static/images/not_image_mime_type.jpg:
```diff
@@ -0,0 +1 @@
+hihi
```
테스트용 파일이라면, 나중에 봐도 알기 쉽게 test용 파일이라는 것을 명시해주는게 좋을 것 같습니다!

aegis1920 replied:
이미 test라는 디렉토리에 담겨져 있어서 따로 test용 파일이라고 명시하지 않았습니다. test디렉토리에 있는 acceptanceTest 패키지와 acceptance 패키지를 합칠 때 acceptance 패키지만 남겨두는 것처럼요! 이것도 팀원과 다 같이 이야기해보면 좋을 것 같네요!

aegis1920 replied:
제가 오해했네요! 파일 안에 있는 내용을 test로 하도록 하겠습니다 :)

ttkmw commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/PostService.java:
```diff
@@ -16,19 +18,32 @@
     public PostResponse findPost(Long id) {
         Post post = findNotDeletedPost(id);
-        return PostResponse.of(post);
+        List<Image> images = imageService.findAllByPostId(id);
```
현재 이미지-포스트는 다대일 관계로 짜주신 것 같습니다!
그러다보니 post를 찾을 때도 post에 해당하는 image들을  찾고나서, 새로운 post를 생성해주고 있는데요!

다대다 관계로 하면 post를 찾는 것만으로도 가능하지 않을까 싶습니다!
물론 ManyToMany를 써다랄고 피드백 드린건 아니고,

Post_Image와 같은 중간 테이블을 둬서 두개의 다대일 관계를 두면 가능하지 않을까 생각해봤습니다!
잘 모르고 드리는 피드백이니 참고만 부탁드려요~

aegis1920 replied:
Post와 Image는 N:N 관계라기보다는 1:N 관계에 더 가까울 것 같은데요! 하나의 Image는 여러 Post와 관계를 갖고 있지 않습니다 :)

ttkmw reviewed (commented):
피드백입니다!
수고하셨습니다~!

ttkmw commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/domain/ImageExtensionTest.java:
```diff
@@ -0,0 +1,24 @@
+
+    @DisplayName("이미지 확장자 유효성 검사")
+    @Test
+    void validateImageExtension_NoneMatchImageExtension_ThrownException() throws IOException {
```
throwException 하지 않는 경우도 테스트하는게 좋지 않을까 싶습니다!

aegis1920 replied:
넵 성공하는 케이스도 만들겠습니다 :)

ttkmw commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/service/ImageServiceTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    @DisplayName("이미지 파일 MIME Type 유효성 검사")
+    @Test
+    void validateImageMimeType_NotImageMimeType_ThrownException() throws IOException {
```
when이 save인데 validate을 테스트하는 것 같습니다

save라고 변경 후, save가 안되는 걸 테스트하는 거라고 설명해주는 게 좋을것같습니다

그리고 save 성공에 대한 테스트도 있는 게 좋을 것 같습니다

aegis1920 replied:
맞습니다! 성공에 대한 테스트도 넣겠습니다 :) 👍

ttkmw commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/repository/ImageRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface ImageRepository extends JpaRepository<Image, Long> {
+
+    List<Image> findAllByPostId(Long id);
```
제 생각엔 이것도 테스트해주면 좋을 것 같긴한데,
개인차인 것 같아 테스트 안해주셔도 괜찮을 것 같긴 합니다 ㅎㅎ

테스트 한번 고려해주세요!

aegis1920 replied:
Spring Data JPA는 자동으로 쿼리를 만들어주는 기능을 가지고 있습니다!
https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#jpa.query-methods
를 참고하면 좋을 것 같아요 :)

ttkmw reviewed (commented):
몇가지 피드백 추가했습니다!

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/aws/S3Uploader.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private void removeNewFile(File targetFile) {
+        // TODO: 2020/07/28 추후 logger 적용
```
이 부분은 Logger 적용해서 올려주셔도 좋을 것 같아요~

aegis1920 replied:
적용했습니다 :)

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/aws/S3Uploader.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private String upload(File uploadFile, String dirName) {
+        String fileName = dirName + "/" + uploadFile.getName();
```
`/`는 상수화해야하는가 고민이네요 😯

aegis1920 replied:
상수화하는 게 더 깔끔할 것 같네요ㅎㅎ

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/aws/S3Uploader.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private Optional<File> convert(MultipartFile file) {
+        File convertFile = new File(Objects.requireNonNull(file.getOriginalFilename()));
```
두 줄로 나눠도 좋을 것 같아요(개인 취향)
나눠보니 안깔끔해보이네요 T-T

``` diff
-         File convertFile = new File(Objects.requireNonNull(file.getOriginalFilename()));
+         Objects.requireNonNull(file.getOriginalFilename())
+         File convertFile = new File(file.getOriginalFilename());
```

세 줄은 너무 긴 느낌이고... 편한대로 해주세요 ㅎㅁㅎ
``` diff
-         File convertFile = new File(Objects.requireNonNull(file.getOriginalFilename()));
+         String originalFileName = file.getOriginalFilename()
+         Objects.requireNonNull(originalFileName)
+         File convertFile = new File(originalFileName);
```

aegis1920 replied:
개인적으로 가독성 차이는 없어보이긴 하지만 두 줄로 나눠도 괜찮을 것 같아 나눠보았습니다 :)

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/GlobalExceptionHandler.java:
```diff
@@ -19,15 +21,17 @@
     @ExceptionHandler(Exception.class)
-    public ResponseEntity<ErrorResponse> handleInternalServerError() {
+    public ResponseEntity<ErrorResponse> handleInternalServerError(Exception e) {
+        e.printStackTrace();
```
제거하거나, 로깅하는게 좋을 것 같아요 :)

aegis1920 replied:
e.getMessage()를 날린다는 게 잘못 썼습니다 ㅠㅠ 아예 제거하면 에러 메세지를 못 보기 때문에 로깅하도록 하겠습니다 :)

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/Image.java:
```diff
@@ -0,0 +1,67 @@
+        return post;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Image image = (Image) o;
+        return Objects.equals(name, image.name) &&
+            Objects.equals(url, image.url) &&
+            Objects.equals(post, image.post);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(name, url, post);
+    }
```
VO가 아닌 엔티티의 경우 equals와 hashcode를 재정의하지 않는 것이 좋아보입니다.

이유) JPA에서의 엔티티는 영속성 컨텍스트를 통해 @Id 필드인 식별자로 구분하는 것이 원칙

aegis1920 replied:
알려주셔서 감사합니다 👍 👍 
https://struberg.wordpress.com/2016/10/15/tostring-equals-and-hashcode-in-jpa-entities/ 에 좋은 자료가 있네요 :)

jamie9504 commented on back-end/legeno-around-here/src/main/resources/logback-spring.xml:
```diff
@@ -1,5 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <configuration scan="true" scanPeriod="60 seconds">
+  <property name="LOG_PATTERN"
+    value="%-5level %d{yy-MM-dd HH:mm:ss}[%thread] [%logger{0}:%line] - %msg%n"/>
+  <property name="LOG_PATH" value="logs/"/>
+  <property name="LOG_FILE_NAME" value="normal_log"/>
+  <property name="ERR_LOG_FILE_NAME" value="err_log"/>
```
수정하신 김에 logging 부분도 넣어주면 좋겠네요~!

aegis1920 replied:
넵 알겠습니다 :)

jamie9504 reviewed (changes requested):
설계 관련 궁금한 건 가서 여쭤볼께요~

생각을 좀 적어봤고, 변경했으면 좋을 것 같은 부분도 있어서 Request Change로 올립니다!

---

설계 관련해서
Post만 Image를 갖는 상황이라면, Image => PostImage로 변경 후,
CASCADE & ORPHAN을 적용하는 게 좋을 것 같아요~!
(Post 객체에서 해당 Image에 `@OneToMany(mappedBy = "post", cascade = CascadeType.ALL + orphanRemoval = true)` 이런식으로요~)

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/dto/PostRequest.java:
```diff
@@ -1,17 +1,32 @@
 import java.util.Objects;
+import org.springframework.web.multipart.MultipartFile;
 import wooteco.team.ittabi.legenoaroundhere.domain.Post;
 
 public class PostRequest {
 
     private String writing;
 
+    private List<MultipartFile> images;
```
공백 제거하면 어떨까요??

aegis1920 replied:
저번 회의 때 DTO나 엔티티 필드에 모두 공백을 넣는 것으로 생각했습니다! 착각했네요ㅠㅠ 수정하겠습니다

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/aws/S3Uploader.java:
```diff
@@ -0,0 +1,67 @@
+        File convertFile = new File(Objects.requireNonNull(file.getOriginalFilename()));
+        try {
+            if (convertFile.createNewFile()) {
+                FileOutputStream fos = new FileOutputStream(convertFile);
+                fos.write(file.getBytes());
```
변수명 약어보다는 fileOutputStream 처럼 쓰는게 좀더 낫지 않을까요?!

aegis1920 replied:
넵 수정하겠습니다 :)

yelimkim98 commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/acceptance/PostAcceptanceTest.java:
```diff
@@ -45,35 +61,43 @@ public void setUp() {
-        assertThat(postResponse.getId()).isEqualTo(id);
-        assertThat(postResponse.getWriting()).isEqualTo(expectedWriting);
+
+        // 이미지가 포함되지 않은 글 등록
+        String postWithoutImageLocation = createPostWithoutImage();
+        Long postWithoutImageId = getIdFromUrl(postWithoutImageLocation);
+        PostResponse postWithoutImageResponse = findPost(postWithoutImageId);
+        assertThat(postWithoutImageResponse.getId()).isEqualTo(postWithoutImageId);
+        assertThat(postWithoutImageResponse.getWriting()).isEqualTo(TEST_WRITING);
+
+        // 이미지가 포함된 글 등록
+        String postWithImageLocation = createPostWithImage();
+        Long postWithImageId = getIdFromUrl(postWithImageLocation);
+        PostResponse postWithImageResponse = findPost(postWithImageId);
+        assertThat(postWithImageResponse.getId()).isEqualTo(postWithImageId);
+        assertThat(postWithImageResponse.getWriting()).isEqualTo(TEST_WRITING);
+        assertThat(postWithImageResponse.getImages()).hasSize(2);
```
`PostResponse postWithoutImageResponse = findPost(postWithoutImageId);` 앞, 뒤에서 enter 를 넣어주면 어떨까요?

글 등록 블럭이 두개인데 private 메서드같은걸로 빼면 어떨까요?

aegis1920 replied:
엔터를 넣어서 가독성을 높였습니다 :) private 메서드는 팀원들과 이야기해보면 좋을 것 같아요!

yelimkim98 reviewed (commented):
1. Post 안에서 Image 의 위치가 어디인지가 저장되지 않고있는 것 같은데, 1차 release 때는 이미지를 항상 Post의 제일 위에 있게 해도 좋을 것 같아요.

다들 어떻게 생각하시는지 궁금합니다.

2. Content 가 쓰일거라고 예상했던 부분에서 안 쓰이고 있네요.
Image 가 Post 를 참조해서 Content 에 Image 를 넣지않는 구조가 된거라면
아얘 Content 객체를 없애버리고
String writing만을 갖는 Writing 객체를 정의하는 것도 좋을 것 같아요.

이것도 다들 어떻게 생각하시는지 궁금하네요.

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/PostController.java:
```diff
@@ -26,7 +26,7 @@ public PostController(PostService postService) {
     @PostMapping
-    public ResponseEntity<Void> createPost(@RequestBody PostRequest postRequest) {
+    public ResponseEntity<Void> createPost(PostRequest postRequest) {
         Long postId = postService.createPost(postRequest).getId();
```
한 줄에 점 하나만 사용하는것이 좋을 것 같습니다!

aegis1920 replied:
융통성있게 `.`을 사용하는 것으로 결정했습니다 👍

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/dto/PostResponse.java:
```diff
@@ -1,31 +1,36 @@
     private Long id;
+
     private String writing;
 
```
12, 14 line의 공백을 띄우신 이유가 있을까요?

aegis1920 replied:
제가 착각했던 부분입니다! 수정하겠습니다 :)

Hamliet commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/service/PostServiceTest.java:
```diff
@@ -2,41 +2,83 @@
+        assertThat(postResponse.getImages()).hasSize(1);
     }
 
     @DisplayName("ID로 포스트 조회 - 성공")
```
포스트 조회가 실패하는 경우도 있지 않을까요? (조회 요청을 했는데 이미 삭제된 경우)

aegis1920 replied:
이미 삭제됐을 경우 실패하는 테스트 케이스도 추가했습니다 :)

Hamliet reviewed (commented):
안녕하세요 라흐입니다. 코드를 보니 정말 많이 노력한 흔적이 보이네요 ^^;
우선 컨벤션 관련한 부분만 피드백 남겼고, 코드 전반적인 설명을 듣고 나서 피드백 할 부분이 있으면 추가하겠습니다!

aegis1920 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#52

aegis1920 committed:
refactor : 루트 피드백 반영

- .jpg 확장자에 있는 내용 수정
- 이미지 저장 성공 테스트 작성

aegis1920 committed:
refactor: logging info 메세지 추가 및 Image 엔티티 equals, hashcode 삭제

aegis1920 committed:
refactor: 이미 지워진 포스트 조회 테스트 추가

- 공백 삭제

aegis1920 committed:
refactor: Image, Post 엔티티에 롬복 적용

ihoph commented:
컨플릭트 수정 반영 해주시면 어프루브 하도록 하겠습니다~!

aegis1920 committed:
refactor: Post, Image 일대다 양방향 관계로 변경

- RequestDTO, ResponseDTO에 롬복 추가
- ImageResponse 추가
- 양방향 관계로 변경한 로직에 따라 테스트 코드도 수정

ttkmw reviewed (approved):
수고하셨슴다~

aegis1920 commented:
> 1. Post 안에서 Image 의 위치가 어디인지가 저장되지 않고있는 것 같은데, 1차 release 때는 이미지를 항상 Post의 제일 위에 있게 해도 좋을 것 같아요.

-> 단방향 관계에서 양방향 관계로 바꾸면서 자연스럽게 `Post` 안에 `List<Image>`가 있게 됐습니다 :)

> 
> 1. Content 가 쓰일거라고 예상했던 부분에서 안 쓰이고 있네요.
>    Image 가 Post 를 참조해서 Content 에 Image 를 넣지않는 구조가 된거라면
>    아얘 Content 객체를 없애버리고
>    String writing만을 갖는 Writing 객체를 정의하는 것도 좋을 것 같아요.
> 
> 이것도 다들 어떻게 생각하시는지 궁금하네요.

-> `Post` 안의 내용들을 `Content` 객체로 만들 예정입니다만 추후 리펙토링 과정에서 해도 괜찮다고 생각이 들었습니다!

aegis1920 committed:
refactor: LoggerFactory -> @Slf4j, FileIOException 추가

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/aws/S3Uploader.java:
```diff
@@ -0,0 +1,72 @@
+
+    public String upload(MultipartFile multipartFile, String dirName) {
+        File uploadFile = convert(multipartFile)
+            .orElseThrow(() -> new IllegalArgumentException("MultipartFile -> File로 전환이 실패했습니다."));
```
해당 예외가 업로드 실패보단 전환 실패에 가까우니 convert에서 해당 예외를 던진 후 File을 줘도 괜찮을 것 같네요!

`IllegalArgumentException`이 아닌 커스텀 익셉션으로 처리해야 Controller에서 핸들링하기 좋을 것 같아요.

aegis1920 replied:
`convert`에서 예외를 던지도록 수정했으며 `MultipartFileConvertException`를 던지도록 했습니다 :)

jamie9504 commented on back-end/legeno-around-here/build.gradle:
```diff
@@ -15,12 +15,15 @@ repositories {
         exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
     }
     testImplementation 'io.rest-assured:rest-assured'
+    compile 'com.amazonaws:aws-java-sdk:1.11.824'
```
https://docs.aws.amazon.com/ko_kr/sdk-for-java/v1/developer-guide/setup-project-gradle.html

오전 의존성을 너무 많이 가져오는 부분에 대해서 위 링크를 참고해보면 좋을 것 같네요!
(이번에 수정할 필요는 없고, 참고요!)

aegis1920 replied:
적용했습니다 :)

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/aws/S3Uploader.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    private void removeNewFile(File targetFile) {
+        if (targetFile.delete()) {
+            log.info("파일이 삭제되었습니다.");
+        } else {
+            log.info("파일이 삭제되지 못했습니다.");
+        }
```
else를 사용하기 전으로 돌아가보는 것은 어떨까요?

``` java
        if (targetFile.delete()) {
            log.info("파일이 삭제되었습니다.");
            return;
        }
        log.info("파일이 삭제되지 못했습니다.");
        
```

aegis1920 replied:
수정했습니다!

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/ImageExtension.java:
```diff
@@ -0,0 +1,44 @@
+    SVG(".svg"),
+    TIF(".tif"),
+    TIFF(".tiff"),
+    WEBP(".webp");
```
BMP, PNG 도 있습니다~!

aegis1920 replied:
웹에서 가능한 이미지만 넣었으며 브라우저에서 제대로 지원하지 않는 확장자(BMP)는 뺐습니다!
https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types를 참고하면 좋을 것 같아요!

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/GlobalExceptionHandler.java:
```diff
@@ -19,17 +24,26 @@
+
     @ExceptionHandler(Exception.class)
-    public ResponseEntity<ErrorResponse> handleInternalServerError() {
+    public ResponseEntity<ErrorResponse> handleInternalServerError(Exception e) {
+        log.info(e.getMessage());
         return ResponseEntity
             .status(HttpStatus.INTERNAL_SERVER_ERROR)
-            .body(new ErrorResponse("예기치 않은 오류가 발생하였습니다."));
+            .body(new ErrorResponse(e.getMessage()));
     }
 }
```
예기치 않은 오류의 경우, 저희가 직접 관리했던 오류가 아닐거라서, getMessage시 어떤 메세지가 전달이 될지 알 수 없습니다.
따라서 기존과 같은 방법으로 일관된 메시지를 넘겨주는 것이 좋다고 생각하는데요!

어떻게 생각하시는지 궁금하네요!

aegis1920 replied:
전에 stash 했던 내용이 그대로 들어갔네요 ㅠ... "예기치 않은 오류가 발생하였습니다."로 다시 넣겠습니다!

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/aws/S3Uploader.java:
```diff
@@ -0,0 +1,72 @@
+        this.bucket = bucket;
+    }
+
+    public String upload(MultipartFile multipartFile, String dirName) {
```
같은 이름의 파일이 s3로 등록됐을 경우의 처리도 해주어야 할 것 같아요~

aegis1920 replied:
보안적인 이슈도 있기 때문에 `posts/images/유저 이메일의 hash값/UUID + 파일명`을 통해 중복 처리를 하도록 결정했습니다!

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/Image.java:
```diff
@@ -0,0 +1,35 @@
+    private String name;
+    private String url;
+
+    @ManyToOne
+    @JoinColumn(name = "post_id")
+    private Post post;
```
Fetch 전략을 Lazy로 설정해도 될 것 같아요

Post는 이미지를 항상 알아야하지만, 이미지의 경우 Post를 항상 알아야하는 경우가 아니니까요~!

aegis1920 replied:
default가 EAGER네요! 조언 감사합니다 👍

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/ImageExtension.java:
```diff
@@ -0,0 +1,44 @@
+    TIF(".tif"),
+    TIFF(".tiff"),
+    WEBP(".webp");
+
```
나중에 이미지 파일을 어디까지 관리할 지 생각해보면 좋을 것 같네요~!
https://itfix.tistory.com/78

aegis1920 replied:
웹에서 가능한 이미지만 넣었으며 브라우저에서 제대로 지원하지 않는 확장자는 뺐습니다!
https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Image_types를 참고하면 좋을 것 같아요!

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/Image.java:
```diff
@@ -0,0 +1,35 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class Image extends BaseEntity {
+
+    private String name;
+    private String url;
```
null을 허용하지 않는다면 

`@Column(nullable = false)`을 설정해둬도 좋을 것 같아요~
물론 DDL 생성시에만 쓰이는 것이긴 하지만, 표시해주면 코드를 볼 때 아 이 엔티티의 해당 필드는 not null이구나! 생각해볼 수 있다고 하더라구요~1

aegis1920 replied:
가독성이 좀 더 좋겠네요! 추가했습니다!

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/dto/ImageResponse.java:
```diff
@@ -0,0 +1,33 @@
+import lombok.ToString;
+import wooteco.team.ittabi.legenoaroundhere.domain.Image;
+
+@NoArgsConstructor
```
모든 DTO의 NoArgsConstructor의 경우 
리플렉션 기능을 이용하기 위함이니, access 범위를 Private으로 지정해주면 좋을 것 같아요 :)

aegis1920 replied:
private으로 지정해줘도 되는군요! 조언 감사합니다 👍

jamie9504 reviewed (changes requested):
리뷰를 조금 더 남겨놓았으니 확인 부탁드려요 :)

Hamliet reviewed (commented):
몇가지 코멘트 남겼습니다.
추가로 아까 말씀드린 서버에서 설정파일을 관리하려면 아래 처럼 수정하시면 됩니다.

`LegenoAroundHereApplication.java` 파일

``` java
public class LegenoAroundHereApplication {

    public static final String APPLICATION_LOCATIONS = "spring.config.location="
        + "classpath:application.properties,"
        + "/app/config/back/profile.yml,"
        + "/app/config/back/[설정 파일 이름]";

    public static void main(String[] args) {
        new SpringApplicationBuilder(LegenoAroundHereApplication.class)
            .properties(APPLICATION_LOCATIONS)
            .run(args);
    }
```

aegis1920 committed:
feat: MultipartFileConvertException 추가

aegis1920 committed:
refactor: s3만 의존하도록 의존성 수정

aegis1920 committed:
refactor: Image @ManyToOne의 fetch 방식을 LAZY로 수정

- ErrorResponse 메시지 수정

aegis1920 committed:
refactor : 가능한 이미지 확장자 추가

aegis1920 committed:
refactor : Post, Image 필드에 nullable 속성 추가

aegis1920 committed:
refactor : NoArgsConstructor의 access level을 private으로 수정

Hamliet reviewed (approved):
고생하셨습니다. 이만 approve할게요! :)

aegis1920 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#52

aegis1920 committed:
feat: MD5 해시 알고리즘 적용한 이미지 파일 중복 처리 및 테스트 수정

- NotFoundAlgorithmException 추가

jamie9504 reviewed (approved):
많은 Comment 반영 잘 해주셨네요!

이만 Approve 합니다.

aegis1920 commented:
추후 master 브랜치에서는 이미지 업로드 기능이 동작해야 합니다. 현재는 모두 Mocking 상태입니다.
